lib/uring/uring.ml

```diff
@@ -349,21 +398,34 @@ let send_msg ?(fds=[]) ?dst t fd buffers user_data =
   let addr = Option.map Sockaddr.of_unix dst in
   let n_fds = List.length fds in
   let msghdr = Msghdr.create_with_addr ~n_fds ~fds ?addr buffers in
-  with_id_full t (fun id -> Uring.submit_send_msg t.uring id fd msghdr) user_data ~extra_data:msghdr
+  (* NOTE: `msghdr` references `buffers`, so it's enough for `extra_data` *)
+  with_id_full t (fun id ->
+      let iovec = Sketch.Iovec.alloc t.sketch buffers in
+      Uring.submit_send_msg t.uring id fd msghdr iovec) user_data ~extra_data:msghdr
 
 let recv_msg t fd msghdr user_data =
-  with_id_full t (fun id -> Uring.submit_recv_msg t.uring id fd msghdr) user_data ~extra_data:msghdr
+  let _, _, buffers = msghdr in
+  (* NOTE: `msghdr` references `buffers`, so it's enough for `extra_data` *)
+  with_id_full t (fun id ->
+      let iovec = Sketch.Iovec.alloc t.sketch buffers in
+      Uring.submit_recv_msg t.uring id fd msghdr iovec) user_data ~extra_data:msghdr
 
 let cancel t job user_data =
   ignore (Heap.ptr job : Uring.id);  (* Check it's still valid *)
   with_id t (fun id -> Uring.submit_cancel t.uring id (Heap.ptr job)) user_data
 
 let submit t =
-  if t.dirty then begin
-    t.dirty <- false;
-    Uring.submit t.uring
-  end else
-    0
+  let v =
+    if t.dirty then begin
+      t.dirty <- false;
+      Uring.submit t.uring
+    end else
+      0
+  in
+  (* We really just need to test if SQEs were consumed in polling mode *)
+  if not t.polling || (Uring.sq_ready t.uring = 0) then
```

Comment:
I was expecting just:
```suggestion
  if Uring.sq_ready t.uring = 0 then
```
i.e. free the buffers if everything has been consumed. If submit only consumes some SQEs then we should just leave it.

If we are polling, I guess we're pretty much certain to have unconsumed SQEs at this point, since we just did a submit without waiting. Might be better to do the release *before* submitting, at least in polling mode. On a really busy server, we might never be idle, but we can fix that later (can just add a TODO somewhere - we're not using polling mode much at the moment).